Add external modules required for aarch64 environment to mozc-deps.yaml

phoepsilonix opened the pull request:
In #28, CPython URLs are compiled in a separate file as they are not expected to change frequently.

In this approach, while the aarch64 version of CPython URL needs to be hardcoded, it is added directly to mozc-deps.yaml.

If a cross-compilation environment were available, we could potentially run bazel build for both x86_64 and aarch64, triggering downloads for both architectures. This might eliminate the need for hardcoding.

```
 I added the URL for the aarch64 version of CPython by hardcoding it within update_mozc_deps and downloading it through a proxy, then adding it to mozc-deps.yaml.
Additionally, I implemented a process to add the 'only-arches' item if the URL contains 'x86_64' or 'aarch64'.
```

PS.
We have switched to updating using qemu and flatpak.
Please check the comment below.

phoepsilonix commented:
flathub/org.fcitx.Fcitx5.Addon.Mozc/pull/13

wengxt commented:
I definitely will not put a hardcoded URL in this script, that would defeat the whole purpose of writing this.

As for "only-arches", I don't really care, since
1. without it it will just waste few MBs for downloading
2. in a better version, we can still try to "intersect" the downloaded file or sth to automatically tag the files, which can comes later.

So far what I'm working on is to launch an virttual aarch environment on x64.

I tried --cpu= , --host_cpu= flag on bazel or set a cross compile gcc to bazel, non of them make bazel to download aarch python.

So I turned to run an aarch bazel instead. This time it seems to help. I tried debian multi arch within docker, with a qemu-static on the host. And it does download aarch python file.

Docker is harder to integrate with the script, so I decide to try the foreign arch with flatpak.

And actually, this does work better:

```
flatpak run --arch=aarch64 --share=network --filesystem=host org.freedesktop.Sdk  # need to install first
```

It can give you an aarch64 environment and it is much more easier and you don't even need to run a bunch of apt to install required packages.

with shared filesystem we could even let the py script to run on host machine natively and it can still help mirror the file. (and reuse the clone'd repo, as long as you set TMPDIR to somewhere sharable.

phoepsilonix commented:
The correct approach would be to use bazel build --nobuild in an environment emulating aarch64, right?
From that perspective, I think emulation using flatpak is the best idea.
You don't have to be conscious of QEMU operations, and file sharing is also easy.
We're almost there.

phoepsilonix commented:
For Arch, it seems that if we install qemu-user-static-binfmt and then install the aarch64 SDK using Flatpak, it should work well. 
Since Bazel in the 23.08 branch has been updated to 7.3.1, building using org.freedesktop.Sdk.Extension.bazel should also be fine.

phoepsilonix commented:
We have switched to updating using qemu and flatpak.

    Added cpython aarch64 to mozc-deps.yaml
    
    To obtain information about necessary external modules in the aarch64 environment,
    we added bazel execution using qemu-user-static-binfmt environment and aarch64 emulation through flatpak.
    
    Additionally, I implemented a process to add the 'only-arches' item if the URL contains 'x86_64' or 'aarch64'.
    
    For execution, the qemu-user-static-binfmt environment and installation of the following flatpak packages are necessary.
    
      org.fcitx.Fcitx5/aarch64/stable
      org.kde.Sdk/aarch64/6.6
      org.freedesktop.Sdk/aarch64/23.08
      org.freedesktop.Sdk.Extension.bazel/aarch64/23.08
      org.freedesktop.Sdk.Extension.llvm18/aarch64/23.08 Note:
      org.kde.Sdk//6.6 and org.freedesktop.Sdk//23.08 are interdependent.
    
    As the bazel version in org.freedesktop.Sdk.Extension.bazel//23.08 has been updated to 7.3.1, it can be used for building with blzmod enabled.
    In the future, it seems that when org.kde.Sdk//6.8 becomes available, org.freedesktop.Sdk//24.08 will be used.

phoepsilonix commented:
@wengxt

I have updated the yq command to conform to the v4.44.3 format.
```sh
yq --version
yq (https://github.com/mikefarah/yq/) version v4.44.3
```

phoepsilonix commented:
I got it.

phoepsilonix commented:
I have prepared a workflow in GitHub Actions to update mozc-deps.yaml.
For now, I will add this file to this branch.
Please check it.

phoepsilonix commented:
I think I've managed to organize the general flow.
The fact that it can be executed in a shorter time than emulating with Docker's arm64 image is a significant advantage, in my opinion.
This workflow has not yet been successfully configured for caching. I believe there's room for improvement.

phoepsilonix commented:
@wengxt

phoepsilonix commented:
@wengxt

I modified update_mozc_deps to switch processing based on the first argument.  
`./update_mozc_deps x86_64` and `./update_mozc_deps aarch64` execute bazel on flatpak. Running only `./update_mozc_deps` executes bazel in the native environment.

I adjusted the workflow cache.
In the workflow, I also tried running bazel with flatpak for x86_64.  
In the workflow, after combining all results, duplicates are removed.

phoepsilonix commented:
#31

phoepsilonix commented:
#32
